File: server-python/src/revturbine/core/placements/threshold_gating.py
"""Python port of scaffold's ``threshold-gating.ts`` shared evaluation.

Byte-faithful port. The TS file is the executable spec — every branch
mirrors a TS branch so cross-language parity stays a true equivalence
check rather than a re-derivation. Keep aligned with the TS when either
side changes.

Usage / credit / seat threshold triggers all gate on percent CONSUMED
(placement-studio-ui.md §3.4):

    usage_threshold   — usage_current / usage_limit * 100
    credit_threshold  — (allocation - balance) / allocation * 100
    seat_threshold    — seats_filled / seat_limit * 100

Credits pass the REMAINING balance, not the consumed amount; RT derives
the percentage. Percent is not clamped to 100 (the "Exceeded" direction).
Thresholds evaluate at the entitlement's Allocation level. Indeterminate
state fails closed.

Plan 138 TASK-3 (parity for the TS gate shipped in scaffold #157/#159).
"""

from __future__ import annotations

import math
from typing import Any, Literal, TypedDict

# ── ThresholdTriggerShape — discriminated union (kind-keyed) ────────────────


class ThresholdTriggerShape(TypedDict):
    kind: Literal["usage_threshold", "credit_threshold", "seat_threshold"]
    entitlement_handle: str
    threshold_percent: float


def _grant_for_allocation(handle: str, state: dict[str, Any]) -> dict[str, Any] | None:
    """Pick the grant whose counters the threshold is measured against, per
    the entitlement's Allocation (§3.4 "Pooling"). Source:
    threshold-gating.ts grantForAllocation.
    """
    grants = state.get("grants")
    if not isinstance(grants, dict):
        return None

    def _lvl(level: str) -> dict[str, Any] | None:
        lv = grants.get(level)
        entry = lv.get(handle) if isinstance(lv, dict) else None
        return entry if isinstance(entry, dict) else None

    declared = _lvl("user") or _lvl("instance") or _lvl("account")
    allocation = declared.get("allocation") if isinstance(declared, dict) else None
    if allocation in ("account_pool", "per_user_pooled"):
        return _lvl("account") or declared
    if allocation == "per_instance":
        return _lvl("instance") or declared
    if allocation == "per_user":
        return _lvl("user") or declared
    return declared


class _Counters(TypedDict, total=False):
    used: float
    limit: float
    remaining: float


def _resolve_counters(handle: str, state: dict[str, Any]) -> _Counters | None:
    """Allocation-scoped grant when one carries a limit, else the flat usage
    map. Source: threshold-gating.ts resolveCounters.
    """
    grant = _grant_for_allocation(handle, state)
    if (
        isinstance(grant, dict)
        and isinstance(grant.get("limit"), (int, float))
        and not isinstance(grant.get("limit"), bool)
    ):
        used = grant.get("used")
        return {
            "used": float(used)
            if isinstance(used, (int, float)) and not isinstance(used, bool)
            else 0.0,
            "limit": float(grant["limit"]),
        }

    usage = state.get("usage")
    entry = usage.get(handle) if isinstance(usage, dict) else None
    if (
        isinstance(entry, dict)
        and isinstance(entry.get("limit"), (int, float))
        and not isinstance(entry.get("limit"), bool)
    ):
        counters: _Counters = {"used": entry.get("used", 0.0), "limit": entry["limit"]}
        if isinstance(entry.get("remaining"), (int, float)) and not isinstance(
            entry.get("remaining"), bool
        ):
            counters["remaining"] = entry["remaining"]
        return counters

    return None


def compute_consumed_percent(
    trigger: ThresholdTriggerShape,
    entitlements: dict[str, Any] | None,
) -> float | None:
    """Percent of the allocation consumed, or ``None`` when it can't be
    determined (no state, no limit, non-positive limit). Not clamped.
    Source: threshold-gating.ts computeConsumedPercent.
    """
    if not isinstance(entitlements, dict):
        return None

    counters = _resolve_counters(trigger["entitlement_handle"], entitlements)
    if counters is None:
        return None
    limit = counters["limit"]
    if not math.isfinite(limit) or limit <= 0:
        return None

    remaining = counters.get("remaining")
    if (
        trigger["kind"] == "credit_threshold"
        and isinstance(remaining, (int, float))
        and not isinstance(remaining, bool)
    ):
        return ((limit - remaining) / limit) * 100

    used = counters["used"]
    if not (isinstance(used, (int, float)) and not isinstance(used, bool) and math.isfinite(used)):
        return None
    return (used / limit) * 100


def matches_threshold_trigger(
    trigger: ThresholdTriggerShape | None,
    entitlements: dict[str, Any] | None,
) -> bool:
    """Non-threshold triggers (``None``) pass through. Fires when consumed
    percent is at or above the configured threshold; fails closed when
    consumption can't be determined. Source:
    threshold-gating.ts matchesThresholdTrigger.
    """
    if trigger is None:
        return True
    consumed = compute_consumed_percent(trigger, entitlements)
    if consumed is None:
        return False
    return consumed >= trigger["threshold_percent"]

File: server-rust/src/placements/threshold_gating.rs
//! Usage / credit / seat threshold eligibility.
//!
//! All three gate on percent **consumed**:
//!
//! - `usage_threshold` — `used / limit * 100`
//! - `credit_threshold` — `(allocation − balance) / allocation * 100`
//! - `seat_threshold` — `seats_filled / seat_limit * 100`
//!
//! Credits pass the **remaining** balance, not the consumed amount; the
//! percentage is derived here. Percent is deliberately **not clamped** to 100 —
//! the "Exceeded" direction is meaningful. Thresholds evaluate at the
//! entitlement's Allocation level.
//!
//! Source: revturbine-scaffold/src/placements/controllers/threshold-gating.ts

use serde_json::Value;

/// A threshold trigger.
#[derive(Debug, Clone, PartialEq)]
pub struct ThresholdTrigger {
    /// `usage_threshold` | `credit_threshold` | `seat_threshold`.
    pub kind: String,
    /// The entitlement whose counters are measured.
    pub entitlement_handle: String,
    /// The percent at or above which the trigger fires.
    pub threshold_percent: f64,
}

/// A finite number field, rejecting booleans (serde declines them already).
fn num(obj: &Value, key: &str) -> Option<f64> {
    obj.get(key)
        .and_then(Value::as_f64)
        .filter(|f| f.is_finite())
}

/// Pick the grant whose counters the threshold is measured against, per the
/// entitlement's Allocation ("Pooling").
///
/// The declared grant is found by precedence (user → instance → account), then
/// its `allocation` redirects to the level that actually owns the counters.
///
/// Source: threshold-gating.ts grantForAllocation
fn grant_for_allocation<'a>(handle: &str, state: &'a Value) -> Option<&'a Value> {
    let grants = state.get("grants")?;
    let at = |level: &str| {
        grants
            .get(level)
            .and_then(|lv| lv.get(handle))
            .filter(|e| e.is_object())
    };

    let declared = at("user")
        .or_else(|| at("instance"))
        .or_else(|| at("account"))?;

    match declared.get("allocation").and_then(Value::as_str) {
        Some("account_pool" | "per_user_pooled") => at("account").or(Some(declared)),
        Some("per_instance") => at("instance").or(Some(declared)),
        Some("per_user") => at("user").or(Some(declared)),
        _ => Some(declared),
    }
}

struct Counters {
    used: f64,
    limit: f64,
    remaining: Option<f64>,
}

/// The allocation-scoped grant when one carries a limit, else the flat usage
/// map.
///
/// Source: threshold-gating.ts resolveCounters
fn resolve_counters(handle: &str, state: &Value) -> Option<Counters> {
    if let Some(grant) = grant_for_allocation(handle, state) {
        if let Some(limit) = num(grant, "limit") {
            return Some(Counters {
                // An absent `used` on a grant that declares a limit means zero
                // consumed, not "indeterminate".
                used: num(grant, "used").unwrap_or(0.0),
                limit,
                remaining: None,
            });
        }
    }

    let entry = state.get("usage")?.get(handle)?;
    let limit = num(entry, "limit")?;
    Some(Counters {
        used: num(entry, "used").unwrap_or(0.0),
        limit,
        remaining: num(entry, "remaining"),
    })
}

/// Percent of the allocation consumed, or `None` when it cannot be determined
/// (no state, no limit, non-positive limit). **Not clamped.**
///
/// Source: threshold-gating.ts computeConsumedPercent
#[must_use]
pub fn compute_consumed_percent(
    trigger: &ThresholdTrigger,
    entitlements: Option<&Value>,
) -> Option<f64> {
    let state = entitlements.filter(|e| e.is_object())?;
    let counters = resolve_counters(&trigger.entitlement_handle, state)?;

    if !counters.limit.is_finite() || counters.limit <= 0.0 {
        return None;
    }

    // Credits report what is LEFT, so consumption is the complement.
    if trigger.kind == "credit_threshold" {
        if let Some(remaining) = counters.remaining {
            return Some(((counters.limit - remaining) / counters.limit) * 100.0);
        }
    }

    if !counters.used.is_finite() {
        return None;
    }
    Some((counters.used / counters.limit) * 100.0)
}

/// Whether a threshold-triggered placement is eligible.
///
/// Non-threshold triggers pass through. Fires when consumed percent is **at or
/// above** the configured threshold, and **fails closed** when consumption
/// cannot be determined — an upgrade prompt driven by unknown usage is worse
/// than none.
///
/// Source: threshold-gating.ts matchesThresholdTrigger
#[must_use]
pub fn matches_threshold_trigger(
    trigger: Option<&ThresholdTrigger>,
    entitlements: Option<&Value>,
) -> bool {
    let Some(trigger) = trigger else {
        return true;
    };
    compute_consumed_percent(trigger, entitlements)
        .is_some_and(|consumed| consumed >= trigger.threshold_percent)
}
